### DISCUSSION TITLE
console output without a newline

### DISCUSSION THREAD

**Discussion by mortax** [winglang/wing#1772] [Q&A] [answered]
`print()` is handy for generating console output -- but it always includes a newline.

is there a way to print console output without a newline?

**Comment by Chriscbr** [comment]
Not at the moment - I think you're the first to bring this to our attention @mortax! This isn't is covered by our [language spec](https://docs.winglang.io/reference/spec) yet.

(2 cents) I feel like there are at least a few directions we could take this...

* Change `print` to not include a newline, and consider a separate `println` function (see Java, Rust, Go)
* Add an optional argument to customize the line ending, as in `print("hello", end: "")` (see Python)

It could also make sense to view `print` as more of a logging functionality -- where each log is its own entry, and we are just displaying each on a newline for convenience. This would align with how `console.log` works in JavaScript and how many cloud logging libraries work.

**Reply by mortax** [reply]
got it. i very much favor the first solution since (_i think_) it's consistent with more other languages than Python's `end:` pattern. the downside, of course, is that neither is backward compatible. that said:
* we're still only in Alpha so better now than later
* i doubt many people are using `print()` at this point (and they can adapt quickly with a simple edit - which could even be automated)
> It could also make sense to view print as more of a logging functionality -- where each log is its own entry, and we are just displaying each on a newline for convenience. This would align with how console.log works in JavaScript and how many cloud logging libraries work.

i completely agree. my first Wing program is a bit of an outlier. console-style logging can still be very helpful, of course. IMHO, i think it would be useful to have the ability to log partial entries.

i did a quick check of my code-base (which covers ~40 languages) and i found nine that use `println`:
* Clojure
* Frink
* Groovy
* Java
* Julia
* Kotlin
* Rust
* Scala
* V

thanks! i will try to work up a PR for further discussion.

**Comment by yoav-steinberg** [comment, marked as answer]
maybe rename `print` to `log`, and see if that's good enough for now. If we ever find that we need a way to manipulate output to the terminal we can add `print`/`println`.

**Reply by Chriscbr** [reply]
Implemented in #1898 ! Let's see how this feels to use 👍

**Comment by MarkMcCulloh** [comment]
A current workaround would be to use [extern](https://docs.winglang.io/reference/spec#52-javascript) to write js that does `process.stdout.write("...")`

**Comment by mortax** [comment]
i finally tried this. not the _prettiest_ but it worked perfectly.

i created `process-helper.js`:

```
exports.write = function(s) {
  return process.stdout.write(s);
};
```

added:

```
class Process {
  extern "./process-helper.js" pub static write(s: str);
} 
```

and use:

```
Process.write("-");
```

thanks!

**Comment by mortax** [comment]
addendum: i [created a PR](https://github.com/winglang/wing/pull/5379) to re-introduce `print()` (sans newline).
